TypeScript_ESLint.snb.md

```diff
@@ -0,0 +1,131 @@
+
+```typescript
+{
+    ForInStatement: (node): void => { ... }
+}
+```
+
+This key will be used to match this rule with the name of an AST node type in the ESLint AST, so that the node argument will always have the type that the rule targets:
+
+https://sourcegraph.com/npm/typescript-eslint/types@540cca215af38ab1a49b6cde8bcb492f4b6e53c5/-/blob/dist/generated/ast-spec.d.ts?L75
+
+## Type-checking
+
+One distinction between TypeScript ESLint and vanilla ESLint is that we have access to the TypeScript type checker and can therefore define rules that make use of type information.
+
+The `for-in-array` rule uses the type checker to determine if we are dealing with an array type:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L23-24
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L32-35
+
+## Dual ASTs
+
+In addition to invoking the type checker, the `for-in-array` rule also makes use of **two ASTs**:
+
+1. the ESTree representation used by ESLint
+2. the TypeScript AST emitted by the TypeScript compiler.
+
+The `node` value passed as an argument to the rule function is from AST #1, an instance of [`TSESTree.Node`](https://sourcegraph.com/npm/typescript-eslint/types/-/blob/dist/generated/ast-spec.d.ts?L641):
+
+The `originalNode` value is from AST #2, an instance of ([`TSNode`](https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@3eab889022c9d1617f275017d6951f663ea57f24/-/blob/packages/typescript-estree/src/ts-estree/ts-nodes.ts?L18:13#tab=references)):
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@HEAD/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L25#tab=def
+
+Why two ASTs instead of one? Because each provides a distinct set of useful functionality.
+
+1. The `TSESTree` representation is used by ESLint internally and enables the plugin to reuse much of the code and linting framework provided by ESLint.
+2. The `TSNode` representation is emitted by the TypeScript compiler and contains type information.
+
+This rule checks if the type of the variable in the expression is an array type using the `TSNode` representation:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@HEAD/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L32-35#tab=def
+
+And then uses the `TSESTree` node to report back results to ESLint:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@HEAD/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L36-39
+
+In order to use both ASTs, the plugin must create both and construct a mapping between the two. The way it does so is by using a co-recursive switch pattern. Let's examine how it does this.
```

Comment:
> co-recursive switch pattern

I do not know what that phrase means, and I have repeatedly written this kind of code before 😂. Can this post define the term, or give a phrase like _"meaning: ..."_?

Vaguely:

 ```suggestion
In order to use both ASTs, the plugin must create both and construct a mapping between the two. 
The way it does so is by using a "co-recursive switch pattern": meaning it creates a recursive function containing a switch statement whose cases may each call that recursive function again.
Let's examine how that looks.
```

(I'm not 100% on that phrasing, but it's the best I've got right now...)